aiohttp/cookiejar.py

```diff
@@ -261,8 +263,30 @@ def filter_cookies(self, request_url: URL = URL()) -> "BaseCookie[str]":
                 request_origin = request_url.origin()
             is_not_secure = request_origin not in self._treat_as_secure_origin
 
+        pairs = []
+        d = hostname
+        while d:
+            p = request_url.path
+            while p:
+                pairs.append((d, p))
+                p = p.rsplit("/", maxsplit=1)[0]
+            pairs.append((d, ""))
+
+            try:
+                d = d.split(".", maxsplit=1)[1]
+            except IndexError:
+                # handle last element for split
+                d = ""
```

Comment:
```suggestion
        while d:
            p = request_url.path
            while p:
                pairs.append((d, p))
                p = p.rpartition("/")[0]
            pairs.append((d, ""))
            d = d.partition(".")[2]
```

I think you could write it like this

Reply:
```python
        pairs: List[Tuple[str, str]] = []
        paths: List[str] = []
        path = request_url.path
        while path:
            paths.append(path)
            path = path.rpartition("/")[0]

        domain_part = hostname
        while domain_part:
            for path in paths:
                pairs.append((domain_part, path))
            pairs.append((domain_part, ""))
            domain_part = domain_part.partition(".")[2]
```

This might be even better if there are lots of domain labels....

Reply:
Haven't benchmarked it, but trying to keep it all within iterators so we don't do multiple splits etc., how about this?
```
paths = itertools.accumulate(request_url.path.split("/"), lambda x,y: f"{x}/{y}")
domains = itertools.accumulate(reversed(hostname.split(".")), lambda x,y: f"{y}.{x}")
pairs = itertools.chain((("", ""),), itertools.product(domains, paths))
```

I think this is the reverse order of what we already have, so the reversed() call below would need to be dropped (and I've merged the append into here too).

Reply:
Will put together a benchmark for all the options

Reply:
time_original: 0.04315150000911672
time_suggestion_with_partition: 0.026855791991692968
time_suggestion_with_partition_minimize_loops: 0.016633624996757135
time_suggestion_with_itertools: 0.020396792009705678

Reply:
```python
import timeit

from yarl import URL
import itertools
from typing import Callable


run_count = 20000
hostname = "www.example.com"
request_url = URL(f"http://{hostname}/this/is/a/typical/url")
pairs = []


def original():
    d = hostname
    while d:
        p = request_url.path
        while p:
            pairs.append((d, p))
            p = p.rsplit("/", maxsplit=1)[0]
        pairs.append((d, ""))
        try:
            d = d.split(".", maxsplit=1)[1]
        except IndexError:
            # handle last element for split
            d = ""


def suggestion_with_partition():
    domain_part = hostname
    while domain_part:
        paths = request_url.path
        while paths:
            pairs.append((domain_part, paths))
            paths = paths.rpartition("/")[0]
        pairs.append((domain_part, ""))
        domain_part = domain_part.partition(".")[2]


def suggestion_with_partition_minimize_loops():
    paths = []
    path = request_url.path
    while path:
        paths.append(path)
        path = path.rpartition("/")[0]

    domain_part = hostname
    while domain_part:
        for path in paths:
            pairs.append((domain_part, path))
        pairs.append((domain_part, ""))
        domain_part = domain_part.partition(".")[2]


def suggestion_with_itertools():
    paths = itertools.accumulate(request_url.path.split("/"), lambda x, y: f"{x}/{y}")
    domains = itertools.accumulate(
        reversed(hostname.split(".")), lambda x, y: f"{y}.{x}"
    )
    pairs = itertools.chain((("", ""),), itertools.product(domains, paths))


def run_suggestion(suggestion: Callable[[], None], times: int) -> float:
    start = timeit.default_timer()
    for _ in range(times):
        pairs.clear()
        suggestion()
    end = timeit.default_timer()
    return end - start


time_original = run_suggestion(original, run_count)
time_suggestion_with_partition = run_suggestion(suggestion_with_partition, run_count)
time_suggestion_with_partition_minimize_loops = run_suggestion(
    suggestion_with_partition_minimize_loops, run_count
)
time_suggestion_with_itertools = run_suggestion(suggestion_with_itertools, run_count)

print(f"time_original: {time_original}")
print(f"time_suggestion_with_partition: {time_suggestion_with_partition}")
print(
    f"time_suggestion_with_partition_minimize_loops: {time_suggestion_with_partition_minimize_loops}"
)
print(f"time_suggestion_with_itertools: {time_suggestion_with_itertools}")
```

Reply:
I updated the suggestion above to `suggestion_with_partition_minimize_loops`

Reply:
I added the following couple of lines into the tests, and, atleast on my machine, itertools comes out marginally better:
```
> python test.py
time_original: 0.14913682299811626
time_suggestion_with_partition: 0.09907035499782069
time_suggestion_with_partition_minimize_loops: 0.07111960600013845
time_suggestion_with_itertools: 0.0670045239967294
> python test.py
time_original: 0.12457079599698773
time_suggestion_with_partition: 0.08952741000393871
time_suggestion_with_partition_minimize_loops: 0.0660165480003343
time_suggestion_with_itertools: 0.0644841980029014
> python test.py
time_original: 0.12488939300237689
time_suggestion_with_partition: 0.09736065800098004
time_suggestion_with_partition_minimize_loops: 0.06708528499439126
time_suggestion_with_itertools: 0.06588387100055115
```

Full script:
```


import timeit

from yarl import URL
import itertools
from typing import Callable


run_count = 20000
hostname = "www.example.com"
request_url = URL(f"http://{hostname}/this/is/a/typical/url")
pairs = []


def original():
    d = hostname
    while d:
        p = request_url.path
        while p:
            pairs.append((d, p))
            p = p.rsplit("/", maxsplit=1)[0]
        pairs.append((d, ""))
        try:
            d = d.split(".", maxsplit=1)[1]
        except IndexError:
            # handle last element for split
            d = ""
    pairs.append(("", ""))
    for p in reversed(pairs):
        ...


def suggestion_with_partition():
    domain_part = hostname
    while domain_part:
        paths = request_url.path
        while paths:
            pairs.append((domain_part, paths))
            paths = paths.rpartition("/")[0]
        pairs.append((domain_part, ""))
        domain_part = domain_part.partition(".")[2]
    pairs.append(("", ""))
    for p in reversed(pairs):
        ...


def suggestion_with_partition_minimize_loops():
    paths = []
    path = request_url.path
    while path:
        paths.append(path)
        path = path.rpartition("/")[0]

    domain_part = hostname
    while domain_part:
        for path in paths:
            pairs.append((domain_part, path))
        pairs.append((domain_part, ""))
        domain_part = domain_part.partition(".")[2]
    pairs.append(("", ""))
    for p in reversed(pairs):
        ...


def suggestion_with_itertools():
    paths = itertools.accumulate(request_url.path.split("/"), lambda x, y: f"{x}/{y}")
    domains = itertools.accumulate(
        reversed(hostname.split(".")), lambda x, y: f"{y}.{x}"
    )
    pairs = itertools.chain((("", ""),), itertools.product(domains, paths))
    for p in pairs:
        ...


def run_suggestion(suggestion: Callable[[], None], times: int) -> float:
    start = timeit.default_timer()
    for _ in range(times):
        pairs.clear()
        suggestion()
    end = timeit.default_timer()
    return end - start


time_original = run_suggestion(original, run_count)
time_suggestion_with_partition = run_suggestion(suggestion_with_partition, run_count)
time_suggestion_with_partition_minimize_loops = run_suggestion(
    suggestion_with_partition_minimize_loops, run_count
)
time_suggestion_with_itertools = run_suggestion(suggestion_with_itertools, run_count)

print(f"time_original: {time_original}")
print(f"time_suggestion_with_partition: {time_suggestion_with_partition}")
print(
    f"time_suggestion_with_partition_minimize_loops: {time_suggestion_with_partition_minimize_loops}"
)
print(f"time_suggestion_with_itertools: {time_suggestion_with_itertools}")


```

Reply:
i.e. The original didn't include the extra append in the other tests, which was already included in itertools with chain().
It also didn't include the iteration which could obviously have been rather different when the itertools one does a lot of the work during iteration, rather than before.

Reply:
```
% python3.12 cookies2.py
time_original: 0.044930542004294693
time_suggestion_with_partition: 0.02851945799193345
time_suggestion_with_partition_minimize_loops: 0.018505583007936366
time_suggestion_with_itertools: 0.026156499996432103
% python3.11 cookies2.py
time_original: 0.06036895800207276
time_suggestion_with_partition: 0.03111395899031777
time_suggestion_with_partition_minimize_loops: 0.01939904199389275
time_suggestion_with_itertools: 0.023035916994558647
% python3.10 cookies2.py
time_original: 0.07786262499575969
time_suggestion_with_partition: 0.037624666991177946
time_suggestion_with_partition_minimize_loops: 0.0272009159962181
time_suggestion_with_itertools: 0.02622854200308211
```

The results seem to be a bit different depending on the python version used

Reply:
Either of the last two are fine with me

Reply:
Interesting, looks like newer versions of Python are becoming more performant with loops or lists maybe. Seems like itertools is fairly consistent.

Reply:
Yeah, I think we should choose between those 2 based on readability/maintainability. Do you think the itertools is preferable? I like that the logic for each part is concise enough to be in a single line personally.

Reply:
I like the compactness of the partial but dislike having to use the lambdas so I'm on the fence and both solutions seem fine to me

Reply:
Yeah, the lambda is probably where it loses some performance. If I take out the lambda it's definitely a bit faster (but, obviously gives the wrong result then).

Reply:
Only way I can think to actually remove the lamba is using re.split() to keep the initial / in the split strings, but regex is a lot slower...

Reply:
I was hoping we could have used `"/".join` as the expression but we need multiple args